crates/driver/src/infra/api/error.rs

```diff
@@ -0,0 +1,104 @@
+use {
+    crate::{
+        domain::competition::{self, quote, solution},
+        infra::api,
+    },
+    hyper::StatusCode,
+    serde::Serialize,
+};
+
+#[derive(Debug, Clone, Copy, Serialize)]
+#[serde(rename_all = "PascalCase")]
+enum Type {
+    QuotingFailed,
+    SolverFailed,
+    SolutionNotFound,
+    DeadlineExceeded,
+    SimulationFailed,
+    Unknown,
+    TransactionPublishingFailed,
+    InvalidAuctionId,
+    MissingSurplusFee,
+    LiquidityError,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Error {
+    error_type: Type,
+    description: &'static str,
+}
+
+impl From<Type> for (StatusCode, axum::Json<Error>) {
+    fn from(value: Type) -> Self {
+        let (description, status_code) = match value {
+            Type::QuotingFailed => ("No valid quote found", StatusCode::BAD_REQUEST),
+            Type::SolverFailed => (
+                "Solver engine returned an invalid response",
+                StatusCode::BAD_REQUEST,
+            ),
+            Type::SolutionNotFound => ("No solution found for given ID", StatusCode::BAD_REQUEST),
+            Type::DeadlineExceeded => ("Exceeded solution deadline", StatusCode::BAD_REQUEST),
+            Type::SimulationFailed => ("Solution simulation failed", StatusCode::BAD_REQUEST),
+            Type::Unknown => ("An unknown error occurred", StatusCode::BAD_REQUEST),
+            Type::TransactionPublishingFailed => (
+                "Failed to publish the settlement transaction",
+                StatusCode::BAD_REQUEST,
+            ),
+            Type::InvalidAuctionId => (
+                "Invalid ID specified in the auction",
+                StatusCode::BAD_REQUEST,
+            ),
+            Type::MissingSurplusFee => (
+                "Auction contains a limit order with no surplus fee",
+                StatusCode::BAD_REQUEST,
+            ),
+            Type::LiquidityError => ("Failed to fetch onchain liquidity", StatusCode::BAD_REQUEST),
```

Comment:
I think this should be a server error of some kind (5xx error). I don't know which one though. Either a generic 500 error, or a 503 (temporarily unavailable because it can't fetch liquidity).

I would probably stick with 500 for simplicity, or use a non-standard 5xx error code that we can make up.

Reply:
I'm not sure the `driver` should even error out due to an error while fetching liquidity. Technically it could be able to find a solution without fetching any additional liquidity (liquidity orders in the auction, perfect CoW, market making).
Also would that error already occur when a single liquidity fetch request fails or only when all of them fail?

Reply:
That is a good point. I think it would make sense for liquidity fetching to never error and instead just return "empty liquidity" in case of an error and just print a `WARN` message. Even Quasimodo, which relies on liquidity, would be able to find CoWs in the absence of any pools.

Reply:
https://github.com/cowprotocol/services/issues/1130